### ISSUE TITLE
Bumping knex to 1.0

### ISSUE THREAD

**Issue by elhigu** [tgriesser/knex#1331]
`Knex` has been around already quite long and there is quite big user space for the current knex APIs already (>100k monthly downloads). 

There has been already couple of rewrites for the implementation of the current API, so there won't be big changes to current internal structure for the current `knex`. 

Next rewrite will make big changes how the library is going to be used in future.

I don't see big issues ahead why we couldn't just bump `knex` version already to stable 1.0.

Good thing with bumping version to 1.0 would be that users of current knex would get finally stable version, which won't introduce backwards incompatible changes every couple of months, bad thing would be that we wouldn't be allowed to break backwards compatibility without changing major version.

One thing I'm not so sure is if we can still keep on specifying parts which are currently undefined behavior. So changing undocumented undefined behavior to documented (and probably different) behavior would be still fine with minor version change? If that would be ok (as I think it would) then most of the future changes would be just to define old undefined behavior to work more consistent way. 

Other way to introduce kind of backward incompatible improvements without jumping to 2.0 would be to make the changes configurable and default to old behavior.

Any thoughts / comments @tgriesser @rhys-vdw @bendrucker @wubzz 

What are changes which needs to be done before putting 1.0 out?

**Comment by wubzz** [comment]
Personally I would like to see the documentation issues addressed before moving onto 1.0. A stable version needs a stable documentation alongside it. Not sure on the status of #1125.

Should also run an overhaul to see what is in the library but not yet documented. Unfortunately there seems to still be some stuff.

**Comment by rhys-vdw** [comment]
> I don't see big issues ahead why we couldn't just bump knex version already to stable 1.0.

Agreed. Might as well do it next release.

> ...bad thing would be that we wouldn't be allowed to break backwards compatibility without changing major version.

Not even sure if this is really a bad thing. We'll be doing the same thing with versions < 1.0, it just wont be as obvious from the version number.

**Comment by rhys-vdw** [comment]
> Personally I would like to see the documentation issues addressed before moving onto 1.0.

What did you have in mind? I completely JSDoc'd Bookshelf recently, and while it improved things dramatically, it was a monumental task.

I worry we'll always have that "next big thing" to include before version 1.0 and we'll end up at 0.30.0 ha ha.

**Comment by wubzz** [comment]
Sorry, I didn't mean to emphasize rewriting and restructring the entire documentation, but rather to make sure that the things that are not well documented, or not documented at all, are.

From [this](https://github.com/tgriesser/knex/labels/bug) filter *(without researching each issue's current status)*, I consider the following to appear fundamental enough to oppose the meaning of **stable**:
- #290
- #524
- #852 & #893 
- #1021 
- #1048 
- #1052 
- #1106 
- #1180 
- #1276 

Now I'm not saying all of these need to be fixed to move onto 1.0.0 -- because let's face it, it's just a number at the end of the day -- but to be considered stable, I personally would see these as blocking issues.

**Comment by tgriesser** [comment]
Yeah, let's hold off a little before 1.0 - I realize it's "just a number", but based on my own use of knex recently I've realized there's a ton of rough patches and missing features I haven't had a chance to address/ticket - and I'd like to do a lot of cleanup before cementing the official 1.0, while breaking as little as possible (hopefully nothing) of current users' behavior.

Here's a list of a few things I've been thinking about that would be good to see before a 1.0:

- Adding alternate transaction syntax:
```js
var trx = knex.transaction()

trx.select('*').from('users').where('id', 1)

// Later:
trx.commit().then(...
```
At first this doesn't look like much of a change but it would allow for transactions to be more lightweight, the "start transaction" would be deferred until the first query execution, which would make them much more usable.

- Using without a pool / providing alternate pool
- Fixing a lot of the issues that have cropped up related to the pool
- Better ergonomics for escaping tables / columns / query re-use
- Tapping in to intercept results at the row level instead of needing to wait for the entire promise. Basically a more robust version of what stream is supposed to achieve. I was looking at the proposed [observable](https://github.com/zenparsing/es-observable) spec here, but I'm open to other ideas, particularly interested in the ideas from the [gtor](https://github.com/kriskowal/gtor)
- Tools which would be helpful in unit / integration testing queries, similar to what [supertest](https://github.com/visionmedia/supertest) does for request/response
- More determinism in generating queries, if we make the underlying construction of queries immutable it'd be possible to fix https://github.com/tgriesser/knex/issues/470 to have proper `.then` A+ semantics
- Coming up with a standard way for adding features in dialects that support them.
- Alternate syntax for update, making it read more like sql: `knex.update(table).set(values)`
- Upsert support where possible

Agreed on documentation, I think I would like to see something of a gitbook or similar, with pages dedicated to getting started, common patterns, etc.

Most of these changes require re-evaluating / cleaning up a lot of the internals and taking another look at the underlying sql libs.

I also realize I've been completely MIA on issues in the last many months - really sorry about that. I am planning on getting back into things very soon (can't give an exact timeline, but should be in the next few weeks - I realize I've also said that a few times, but I do actually mean that now). But I agree with @elhigu and others I definitely would like to get this on track for 1.0 asap.

**Comment by elhigu** [comment]
@tgriesser Most of the things in your list for 1.0 seem to be additional features, which could be added also after 1.0 is already released since they shouldn't break backwards compatibility. Also testing tools for users of knex may be added later or be provided as separate modules etc.

Would be good to try to isolate the most-most important breaking features which we would like to get in before stable. For me the most important would be to make current functionality more robust, documented and consistent and to test all APIs that they actually take arrays / functions / raw as parameters as expected (e.g. orderBy with multiple columns each with separate directions doesn't work, or passing knex.raw() to select etc.).

Also as @wubzz mentioned, we should go through all 'bug' issues and fix them 👍 

About the making streams better I would vote use of observables. I've been using bacon / rxjs quite a lot lately and they seem to be more widely used than gtor (specially rxjs has equivalent lib for pretty much every language).

**Comment by tgriesser** [comment]
> Most of the things in your list for 1.0 seem to be additional features, which could be added also after 1.0 is already released since they shouldn't break backwards compatibility. 

Yes, but before a 1.0 gets released I would like to overhaul a lot of how the internals are setup. There will be some minor breaking changes for 1.0, but there will be alpha & beta releases before this happens, along with well defined warning messages for what has changed and why.

Unfortunately I haven't had as much time to spend on getting what I've worked on locally up for others to look at, since it splits the codebase into ~40 or so independent modules and isn't quite finished and needs a bit of context for explanation.

> Also testing tools for users of knex may be added later or be provided as separate modules etc.

Testing currently is very hacked together and ugly. There isn't a standard well thought out lower level API for dealing with different dialect adapters, there is no well designed way of logging queries, there's no way of executing multiple queries on a single connection outside of a transaction. In order to build something I'd be comfortable putting a 1.0 on, there's a lot of work that still needs to be done around these sorts of things.

> and they seem to be more widely used than gtor

As I understood it, the [gtor](https://github.com/kriskowal/gtor) isn't a library, it's just a well written explanation of the different functional abstractions utilized by rxjs, etc. and how they fit together.

**Comment by elhigu** [comment]
Ah. You are right gtor is just an explanation of many different concepts, I thought it had own APIs when I skimmed that through. 

Would be great to get some insight what is coming before it happens and maybe being able to help. Could that local work be shared in some separate knex branch?

**Comment by tgriesser** [comment]
Closed in favor of discussion on #1659

**Comment by sanket143** [comment]
> I worry we'll always have that "next big thing" to include before version 1.0 and we'll end up at 0.30.0 ha ha.

uhmm... well.. [v0.95.10](https://github.com/knex/knex/releases/tag/0.95.10) 😶

**Comment by kibertoad** [comment]
1.0.0 is coming this year. Pinky promise.

**Comment by ibratoev** [comment]
When is this year? :)

**Comment by kibertoad** [comment]
@ibratoev @OlivierCavadenti Got delayed, since we want to also get https://github.com/knex/knex/pull/4471 in, as it is a breaking change as well. After that is in, I will cut a release immediately.

**Comment by OlivierCavadenti** [comment]
> @ibratoev @OlivierCavadenti Got delayed, since we want to also get #4471 in, as it is a breaking change as well. After that is in, I will cut a release immediately.

Yeah I will finish #4471 today.
